### PR TITLE
PrimaryIdentifier microservice enhancements (and misc typo fixes)

### PR DESCRIPTION
**Pull request by vladimir-mencl-eresearch**
### All Submissions:

* [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [X] Have you added an explanation of what problem you are trying to solve with this PR?
* [X] Have you added information on what your changes do and why you chose this as your solution?
* [No] Have you written new tests for your changes?
* [N/A] Does your submission pass tests?
* [X] This project follows PEP8 style guide. Have you run your code against the 'flake8' linter?

Hi,

I'm deploying SATOSA with an OpenIDConnect front-end and SAML2 backend - to allow connecting OpenIDConnect services into our SAML federation (Tuakiri).

As some of our IdPs use eduPersonTargetedID while most uses SAML2 Persistent NameID, I need the PrimaryIdentifier MicroService to allow pulling the right identifier (either EPTID or NameID) from what the SAML2 backend received.

However, I ran into several issues:
* the way attribute values were pulled based on attribute name, `name_id` could not be used - fixed in cba6cd0
* I needed a way to replace the `subject_id` with the value constructed by the PrimaryIdentifier MicroService.  Even though this would ideally be configured via `user_id_from_attrs` in `internal_attributes.yaml`, I could not use this path - as `_auth_resp_callback_func` in `satosa/base.py` processes `user_id_from_attrs` BEFORE calling ResponseMicroServices; the PrimaryIdentifier MicroService comes in too late.  Hence I added a new option `replace_subject_id` for this MicroService in 4d3020e
* while in the code, I saw that the `clear_input_attributes` functionality would not work, so I included a fix for that as well ( 8ae4696 )
* And I've fixed some typos/omissions in the example files and documentation.

Please let me know if any of this needs further clarification or improvement of this is OK to merge.

Thanks a lot in advance for getting back to me.

Cheers,
Vlad

### CODE REVIEW

File: src/satosa/micro_services/primary_identifier.py [hunk 1]
```diff
@@ -54,7 +54,7 @@ def constructPrimaryIdentifier(self, data, ordered_identifier_candidates):
 
             # Get the values asserted by the IdP for the configured list of attribute names for this candidate
             # and substitute None if the IdP did not assert any value for a configured attribute.
-            values = [ attributes.get(attribute_name, [None])[0] for attribute_name in candidate['attribute_names'] ]
+            values = [ attributes.get(attribute_name, [None])[0] for attribute_name in candidate['attribute_names'] if attribute_name != 'name_id' ]
```
Comment (c00kiemon5ter) [inline src/satosa/micro_services/primary_identifier.py:57]: Please, correct me if I'm wrong.


The final value is a concatenation of the attribute-values of the candidate attributes. The candidates are a list of fallback-sets. `None` coming from `name_id` should not be an issue, as the configuration defines that _it is needed_ - if it is not the configuration should include all cases for the expected attributes.

The policy below says that if eptid is there, it should be used in combination with eppn. If it is not, then use eppn with NameID. Otherwise use eppn or NameID alone.

```yaml
ordered_identifier_candidates:
  - attribute_names:
      - eppn
      - eptid
  - attribute_names:
      - eppn
      - name_id
    name_id_format: urn:oasis:names:tc:SAML:2.0:nameid-format:persistent
  - attribute_names:
      - eppn
  - attribute_names:
      - name_id
    name_id_format: urn:oasis:names:tc:SAML:2.0:nameid-format:persistent
```

This means that:
- the combination of eppn&eptid is preferred (final value `<eppn><eptid>[<@scope>]`)
- if that is not available (eppn or eptid or both are not defined), then use the combination eppn&NameID when the name_id format is persistent (final value `<eppn><NameID>[<@scope>]`)
- if that is not available (eppn or name-id or both are not defined, or the name-id-format differs), then use eppn alone (final value `<eppn>[<@scope>]`)
- if that is not available (eppn is not defined), then use the name-id alone (final value `<NameID>[<@scope>]`)

---

In other words, all `attribute_names` are mandatory to have a value. With this change we say that we make an exception for name_id and it's now _optional_. This may make some use-cases easier to define (less fallback sets) but it will change the meaning of existing configurations and possibly break them (the cases that needed to match only when name_id had a value).

---

- Is this a configuration issue? and does the above help with setting up your configuration correctly?
- Have I missed something that should be addressed?

Let me know, I'm open to suggestions.

Comment (vladimir-mencl-eresearch) [reply src/satosa/micro_services/primary_identifier.py:57]: Hi,

Thanks for the detailed review.

I agree that all `attribute_names` should be mandatory - and I'll try explaining why that still holds.

The key to that is that `name_id` is handled separately, around line 65 - in the block headed by `if 'name_id' in candidate['attribute_names']:`

If a matching NameID is found (a value is provided, and the required format is specified and matches the format actually provided), than the value (stored in `candidate_nameid_value`) is added into `values`.

On the other hand, if `name_id` is specified in the candidate's `attribute_names`, but a suitable value is not found, `candidate_nameid_value` is set to `None` - and this value is still added into `values`, invalidating the set produced by this `candidate` entry.  So the invariant is preserved.

The motivation for this change is that the expression this discussion is focused on (L57), when `candidate['attribute_names']` contains `name_id`, would *always* including a `None` in `values`, immediately marking it as invalid - because no value would ever be found.  Values for `name_id` are not present in `attributes`, they're only available as `data.subject_id` (value) + `data.subject_type` (NameIDFormat).

As this code already has special handling for `name_id` attribute names (as per above), this change only complements the special handling by excluding the attribute from the regular handling.

However, a `name_id` value is still mandatory - this is taken care of in the special handling block.

Hope this explains it.

Cheers,
Vlad

File: src/satosa/micro_services/primary_identifier.py [hunk 2]
```diff
@@ -120,6 +120,18 @@ def constructPrimaryIdentifier(self, data, ordered_identifier_candidates):
...
+            # Clear input attributes if so configured.
+            if clear_input_attributes:
+                attributes_to_clear = [attribute_name for attribute_name in candidate['attribute_names'] if attribute_name != 'name_id']
+                msg = "{} Clearing values for these input attributes: {}".format(
+                    logprefix, attributes_to_clear
+                )
+                logline = lu.LOG_FMT.format(id=lu.get_session_id(context.state), message=msg)
+                logger.debug(logline)
+                for attribute in attributes_to_clear:
+                    del data.attributes[attribute]
```
[the remark this answers is not in the document]
Comment (vladimir-mencl-eresearch) [reply src/satosa/micro_services/primary_identifier.py:133]: Thanks - I wasn't sure what the intention was.

If it is to clear all values, all fine with me.  I don't actually even need it for my use case, I just saw it was broken.

I'll revert my commit changing the clear_attributes handling - and push in another one fixing what I saw was broken.

File: src/satosa/micro_services/primary_identifier.py [hunk 3]
```diff
@@ -191,6 +203,12 @@ def process(self, context, data):
                 clear_input_attributes = self.config['clear_input_attributes']
             else:
                 clear_input_attributes = False
+            if 'replace_subject_id' in config:
+                replace_subject_id = config['replace_subject_id']
+            elif 'clear_input_attributes' in self.config:
```
[the remark this answers is not in the document]
Comment (vladimir-mencl-eresearch) [reply src/satosa/micro_services/primary_identifier.py]: Thanks, that was indeed a wrong copy/paste - thanks for catching that!

File: src/satosa/micro_services/primary_identifier.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vladimir-mencl-eresearch) [reply src/satosa/micro_services/primary_identifier.py]: I've now reverted the original `clear_input_attributes` change ( a0b4859 ) and pushed in a minor fix for where it was breaking ( bd98c3d ).

Please let me know how this PR looks to you now.

Cheers,
Vlad

File: src/satosa/micro_services/primary_identifier.py (hunk 1 above)
Comment (c00kiemon5ter) [reply src/satosa/micro_services/primary_identifier.py:57]: yes, you are right! thanks for the explanation

File: src/satosa/micro_services/primary_identifier.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (c00kiemon5ter) [reply src/satosa/micro_services/primary_identifier.py]: looks good!
